internal/controller/oauth/client_credentials_token_provider.go

```diff
@@ -0,0 +1,87 @@
+// Copyright Envoy AI Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package oauth
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ClientCredentialsTokenProvider implements the standard OAuth2 client credentials flow.
+type ClientCredentialsTokenProvider struct {
+	tokenSource oauth2.TokenSource
+	client      client.Client
+}
+
+// NewClientCredentialsProvider creates a new client credentials provider.
+func NewClientCredentialsProvider(cl client.Client) *ClientCredentialsTokenProvider {
```

Comment:
also i don't see any reason why this is exported. Just initialize the token provider inside `NewOIDCProvider`. It's an implementation detail and no reason to leak it to controller. In fact it's only used in `oidcProvider := oauth.NewOIDCProvider(oauth.NewClientCredentialsProvider(b.client), oidc)`.

Reply:
ClientCredentialProvider is the only supported token source currently, if we support other token source in future then we need to export it.